### ISSUE TITLE
Should SSR be on by default?

### ISSUE THREAD

**Issue by probablyup** [styled-components/babel-plugin-styled-components#82]
I was thinking about it and it seems like this would be a sensible thing to enable by default.

**Comment by mxstbr** [comment]
How so? I think most people are using this plugin to get the minification and the nicer class names, I doubt a majority of them do SSR?

**Comment by probablyup** [comment]
That's probably true, but I guess what would be the downside?

Off the top of my head, it seems like it would be useful for testing. Especially if people use snapshot tests, since the hash should stay consistent?

**Comment by jaydenseric** [comment]
SSR is a best practice; tools should make SSR as easy as possible and promote it as the default approach. Opt-out instead of opt-in.

**Comment by mxstbr** [comment]
> SSR is a best practice; tools should make SSR as easy as possible and promote it as the default approach. Opt-out instead of opt-in.

I disagree, I'm fairly certain many people who use React don't do server-side rendering. It'd not a best practice at all, for some apps and use cases it's a must but for some it's totally unnecessary.

**Comment by jaydenseric** [comment]
> many people who use React don't do server-side rendering

Best practice is rarely common practice. For example, most websites are not built with styled-components or even React 😉

For a user there are only upsides and no downsides to a SSR experience. Refresh a page with and without it and you can immediately see the difference. For many commercial applications like ecommerce and marketing sites SSR is not just nice but a requirement. Quality auditing tools like Google's Lighthouse check for it, so it's not just my opinion.

SSR in the React world is not as common as it should be because until recently the tools have not made it easy. I took time off work and spent almost 3 months once trying to get an app to work isomorphically, back when I was learning the React ecosystem at the height of the JS fatigue era. Look how popular Next.js became overnight; it solved a real problem for a lot of people.

**Comment by mxstbr** [comment]
> For a user there are only upsides and no downsides to a SSR experience.

That's not true. As with everything, there's tradeoffs to doing SSR:

- Slower first paint because the server has to do more work and you can't render loading indicators
- The server has to do more work, which might lead to instability
- Until the JS bundle has loaded and rehydrated the entire app is unresponsive to user input
- User input might go missing when the app rehydrates (imagine a form where a user starts typing before the JS bundle has loaded)
- ...

> For many commercial applications like ecommerce and marketing sites SSR is not just nice but a requirement.

I run a production application myself that relies on SSR heavily, and I'm very well aware that it's a requirement for some categories of applications. (I never said anything else?)

Still, I'm pretty sure that  less than 50% of apps out there which are built with React leverage server-side rendering. It might very well be even below 5% or 10%—although we'll never know for sure.

----

All of that being said, it's irrelevant anyway—I get @probablyup's point now which is that enabling the SSR option by default doesn't hurt people who don't do SSR, so why make it an option at all. That's a good point, maybe we should enable it by default!
